### PR TITLE
(#259) use better way to assign id for LLVM

### LINKED ISSUE #259: codegen: a better way to generate valid LLVM IR
Generate IDs for temporary variables after all instructions generated.

To reach the target, I believe a possible solution was replacing `u64` with `Rc<ID>`. Therefore we could assign IDs later and each reference to it was not outdated. And since our `instructions` is a simple sequence, assign IDs after all instructions generated are possible done in the current code structure.

### PR DESCRIPTION
**Pull request by dannypsnl**
resolve #259

### CODE REVIEW

**Comment by dannypsnl** [comment]
@YanHerChen There has some trouble in my local editor setup, therefore, the formatter didn't work as expected, don't merge before I fixed it.

**Comment by dannypsnl** [comment]
The problem was `cargo fmt` would only format `main.rs` and `lib.rs`:
```
$ cargo fmt -v
[lib (2018)] "/Users/routedan/workspace/dannypsnl/elz/src/lib.rs"
[bin (2018)] "/Users/routedan/workspace/dannypsnl/elz/src/main.rs"
rustfmt --edition 2018 /Users/routedan/workspace/dannypsnl/elz/src/lib.rs /Users/routedan/workspace/dannypsnl/elz/src/main.rs
```

**Comment by dannypsnl** [comment]
@YanHerChen Seems like it caused by nested block is too deep for `rustfmt`, I have no idea how to fix this. Just keep in mind and format it manually now.

**Comment by dannypsnl** [comment]
@YanHerChen If looks good to you, then **Approve** and **Merge** this.

File: src/codegen/ir.rs [hunk 1]
```diff
@@ -181,7 +217,7 @@ impl Body {
         use ast::ExprVariant::*;
         match &expr.value {
             Binary(lhs, rhs, op) => {
-                let id = self.get_and_update_count();
+                let id= ID::new();
```
Comment (dannypsnl) [inline src/codegen/ir.rs]: formatting
```suggestion
                let id = ID::new();
```

File: src/codegen/ir.rs [hunk 2]
```diff
@@ -190,7 +226,7 @@ impl Body {
                 }
                 .to_string();
                 let inst = Instruction::BinaryOperation {
-                    id,
+                    id:  id.clone(),
```
Comment (dannypsnl) [inline src/codegen/ir.rs]: formatting
```suggestion
                    id: id.clone(),
```

**Review by YanHerChen — APPROVED** [review]

**Review by YanHerChen — APPROVED** [review]

**Review by YanHerChen — APPROVED** [review]

File: src/codegen/ir.rs [hunk 3]
```diff
@@ -59,15 +61,36 @@ pub(crate) struct TypeDefinition {
...
 pub(crate) struct Label {
-    pub(crate) id: u64,
+    pub(crate) id: Rc<RefCell<ID>>,
 }
 
 impl Label {
-    pub(crate) fn new(id: u64) -> Label {
-        Label { id }
+    pub(crate) fn new(id: Rc<RefCell<ID>>) -> Rc<Label> {
+        Rc::new(Label { id })
```
Comment (YanHerChen) [inline src/codegen/ir.rs:93]: why replaces u64 by Rc<RefCell<ID>>

File: src/codegen/ir.rs (hunk 2 above)
Comment (YanHerChen) [inline src/codegen/ir.rs]: Formatting

id:  id.clone(),
to
id: id.clone(),

File: src/codegen/ir.rs [hunk 4]
```diff
@@ -210,15 +246,15 @@ impl Body {
...
                         let inst = Instruction::FunctionCall{
-                            id,
+                            id: id.clone(),
                             func_name: format!("@{}", name),
                             ret_type: ret_type.clone().into(),
                             args_expr,
                         };
                         self.instructions.push(inst);
-                        Expr::id(ret_type.clone(),id)
+                        Expr::id(ret_type.clone(),    id)
```
Comment (YanHerChen) [inline src/codegen/ir.rs]: why we need four blank?

File: src/codegen/llvm.rs [hunk 5]
```diff
@@ -77,7 +77,7 @@ impl LLVMValue for ir::Instruction {
                 s.push_str(
                     format!(
                         "%{} = {} {} {}, {}",
-                        id,
+                        id.borrow(),
                         op_name,
```
Comment (YanHerChen) [inline src/codegen/llvm.rs:81]: What are the Benefits of borrow()?

Comment (dannypsnl) [reply src/codegen/llvm.rs:81]: This is not a benefit but required. `Rc<RefCell<ID>>` cannot use as `ID`, and `.borrow()` would take out `ID` from it.

File: src/codegen/ir.rs (hunk 3 above)
Comment (dannypsnl) [reply src/codegen/ir.rs:93]: The target is we want to assign `u64` value after instructions generated, therefore we should provide `Rc` to make it able to refer to the same `ID` and update `ID` later.

**Pull request merged**